rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -389,14 +390,44 @@ object Reduce {
     private def evalExplicit(neu: New)(env: Env[Par], rand: Blake2b512Random): M[Unit] =
       eval(neu)(env, rand)
     private def eval(neu: New)(implicit env: Env[Par], rand: Blake2b512Random): M[Unit] = {
-      def alloc(level: Int): Env[Par] =
-        (env /: (0 until level).toList) { (_env, _) =>
+      def alloc(count: Int, urns: Seq[String]): M[Env[Par]] = {
+        val simpleNews = (0 until (count - urns.size)).toList.foldLeft(env) { (_env, _) =>
           val addr: Par = GPrivate(ByteString.copyFrom(rand.next()))
           _env.put(addr)
         }
+        def addUrn(urn: String, newEnv: Env[Par]): Either[ReduceError, Env[Par]] =
+          urnMap.get(urn) match {
+            case Some(p) => Right(newEnv.put(p))
+            case None    => Left(ReduceError(s"Unknown urn for new: ${urn}"))
+          }
+        // This is non-functional, but it's a self-contained implementation of a
+        // functional early terminating left fold, and the library does the same
+        // thing with vars.
+        def foldLeftEarly[A, B, E](as: Seq[A],
+                                   init: B,
+                                   cata: (A, B) => Either[E, B]): Either[E, B] = {
+          var result: Either[E, B] = Right(init)
```

Comment:
I think the body of this method can be replaced with:
`Foldable[List].foldM[Either[E, ?], A, B](as.toList, init)((x, y) => cata(y, x))`

Please replace it and remove the comment on the code being non-functional :)

We could even use the foldM directly, but then implicit resolution starts fussing and we end up in the 'apologizing' part of FP in Scala.

After we have `SeqInstances.traverseSeq` available from a util module, we can drop the `.toList` - please add a TODO comment for that?

Reply:
One more reason scala sucks: I would have never expected foldM to short circuit.